src/resmom/mom_comm.c

```diff
@@ -3941,6 +3954,250 @@ im_request(int stream, int version)
+				break;
+			}
+
+			if( version == IM_OLD_PROTOCOL_VER) {
+				/*
+				 * The arg list is ended by an empty (zero length)
+				 * string.
+				 */
+				num = 4;
+				argv = (char **)calloc(sizeof(char *), num);
+				assert(argv);
+				for (i=0;; i++) {
+					if ((cp = disrst(stream, &ret)) == NULL)
+						break;
+					if (ret != DIS_SUCCESS)
+						break;
+					if (*cp == '\0') {
+						/* got a empty string, end of args lits */
+						free(cp);
+						break;
+					}
+					if (i == num-1) {
+						num *= 2;
+						argv = (char **)realloc(argv,
+						num*sizeof(char *));
+						assert(argv);
+					}
+					argv[i] = cp;
+				}
+			} else {
+			  	argc = disrui(stream, &ret);
+				if (ret != DIS_SUCCESS) {
+					sprintf(log_buffer, "SPAWN_MULTI read of argc");
+					free(node_list);
+					goto err;
+				}
+				argv = (char **)calloc(argc+1, sizeof(char *));
+				assert(argv);
+				for (i=0; i<argc; i++) {
+					argv[i] = disrst(stream, &ret);
+					if (ret != DIS_SUCCESS)
+						break;
+				}
+			}
+			argv[i] = NULL;
+			if (ret != DIS_SUCCESS) {
+				arrayfree(argv);
+				free(node_list);
+				sprintf(log_buffer, "SPAWN_MULTI read of argv array");
+				goto err;
+			}
+
+			num = 8;
+			envp = (char **)calloc(sizeof(char *), num);
+			assert(envp);
+			for (i=0;; i++) {
+				if ((cp = disrst(stream, &ret)) == NULL)
+					break;
+				if (ret != DIS_SUCCESS)
+					break;
+				if (*cp == '\0') {
+					free(cp);
+					break;
+				}
+				if (i == num-1) {
+					num *= 2;
+					envp = (char **)realloc(envp,
+						num*sizeof(char *));
+					assert(envp);
+				}
+				envp[i] = cp;
+			}
+			envp[i] = NULL;
+			if (ret != DIS_EOD) {
+				arrayfree(argv);
+				arrayfree(envp);
+				free(node_list);
+				sprintf(log_buffer, "SPAWN_MULTI read of envp array");
+				goto err;
+			}
+
+			/* Allocate a temporary place to store task IDs */
+                	taskidlist = (tm_task_id *)calloc(list_size, sizeof(tm_task_id));
+			assert(taskidlist);
+
+			/*
+			** Go through the list of nodes and do something
+			** when it's me, otherwise keep looking
+			*/
+			for (i = 0; i < list_size; i++) {
+				if (pjob->ji_nodeid != TO_PHYNODE(node_list[i])) {
+					continue;
+				}
+					
+#ifdef PMIX
+				pbs_pmix_register_client(pjob, node_list[i], &envp);
+#endif
+				if ((ptask = momtask_create(pjob)) == NULL) {
+					sprintf(log_buffer, "momtask_create "
+						"failed to create a new task\n");
+					log_event(PBSEVENT_DEBUG3,
+						PBS_EVENTCLASS_JOB, LOG_DEBUG,
+						jobid, log_buffer);
+					break;
+				}
+				strcpy(ptask->ti_qs.ti_parentjobid, jobid);
+				ptask->ti_qs.ti_parentnode = pvnodeid;
+				ptask->ti_qs.ti_myvnode    = node_list[i];
+				ptask->ti_qs.ti_parenttask = fromtask;
+				if (task_save(ptask) == -1) {
+					sprintf(log_buffer, "unable to save "
+						"task info\n");
+					log_event(PBSEVENT_DEBUG3,
+						PBS_EVENTCLASS_JOB, LOG_DEBUG,
+						jobid, log_buffer);
+					break;
+				}
+				errcode = start_process(ptask, argv, envp, false);
```

Comment:
I think this is the culprit!! Can we do the work of starting the process AFTER we have ack'd (sent IM_ALL_OKAY)?  
Of course, we have a small risk that start_process() actually fails later, even though we would have ack'd okay, but that is as good as the process starting and failing immediately (or similar). In fact, we do the same for job starts in the mom anyway, we ack the send from the server server before actually starting the job/task. So, it does make sense in a parallel spawn to be asynchronous in that way.